### PR TITLE
Enc, Dec: support variant

### PR DESCRIPTION
**Pull request by drewdzzz**
The commit adds variant support to encoder and decoder.

### CODE REVIEW

File: src/mpp/Dec.hpp [hunk 1]
```diff
@@ -239,7 +259,7 @@ struct Resolver {
 	static constexpr size_t SIZE = path_item_static_size(PI);
 	static constexpr size_t USR_ARG_COUNT = path_item_static_size(P0);
 	static_assert(path_item_type(P0) == PIT_STATIC_L0);
-	static_assert((is_path_item_static(PI) && POS < SIZE) ||
+	static_assert(((is_path_item_static(PI) || path_item_type(PI) == PIT_VARIANT) && POS < SIZE) ||
 		      (!is_path_item_static(PI) && POS + SIZE == 0));
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:263]: This assert becomes overly complicated, and it's hard to understand. Could you please leave a comment or add a helper?

File: src/mpp/Dec.hpp [hunk 2]
```diff
@@ -1167,6 +1189,37 @@ bool jump_read_optional(BUF& buf, T... t)
...
+	static_assert(I < size);
+	using curType = std::variant_alternative_t<I, variant_t>;
+	constexpr auto curFamilies = detectFamily<BUF, curType>();
+
+	if constexpr (family_sequence_contains<FAMILY>(curFamilies)) {
+		variant.template emplace<I>();
+		using NEXT_PATH = path_push_t<PATH, PIT_VARIANT, size, I>;
+		return decode_impl<NEXT_PATH>(buf, t...);
+	} else {
+		return jump_read_variant_impl<I + 1, FAMILY, SUBRULE, PATH>(buf, t...);
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:1209]: Shouldn't `variant::emplace` encapsulate this matching? See overload (1) from [std::variant](https://en.cppreference.com/w/cpp/utility/variant/emplace).

**Review by CuriousGeorgiy — APPROVED** [review]

**Comment by CuriousGeorgiy** [comment]
@drewdzzz

BTW, I believe you forgot to support the [`std::monostate`](https://en.cppreference.com/w/cpp/utility/variant/monostate) case, though I guess right now using `optional` will work — please open a ticket.

File: src/mpp/Dec.hpp (hunk 2 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp:1209]: Do you mean matching from `mpp::FAMILY` to variant alternative?

Comment (CuriousGeorgiy) [reply src/mpp/Dec.hpp:1209]: Yes. Can't we make it work like `variant.template emplace<FAMILY>()`?

**Comment by drewdzzz** [comment]
I filed an issue about `std::monostate`, but now `std::nullptr_t` can perform the same thing.

File: src/mpp/Dec.hpp (hunk 2 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp:1209]: How? `std::variant` knows nothing about our msgpack type families.

File: src/mpp/Dec.hpp (hunk 1 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp]: Clarified the assertion.

File: src/Utils/Traits.hpp [hunk 3]
```diff
@@ -554,6 +555,55 @@ constexpr bool is_variant_v =
...
+};
+};
+
+/**
+ * Invokes the functor on current alternative of variant.
+ */
+template <class F, class V>
+void visit(F &&f, V &&v)
+{
+	details::VisitJumps<F, V> jumps;
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/Utils/Traits.hpp]: Oh, thank you!
Fixed.

File: src/mpp/Constants.hpp [hunk 4]
```diff
@@ -209,6 +209,54 @@ static constexpr auto family_sequence_populate(struct family_sequence<FAMILY...>
 	return family_sequence<NEW_FAMILY, FAMILY...>{};
 }
 
+template <compact::Family ...FAMILY_A, compact::Family ...FAMILY_B>
+static constexpr auto family_sequence_concat(struct family_sequence<FAMILY_A...>,
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/mpp/Constants.hpp]: Agree.
Rewrote it as `operator+` without `struct` and `static` keywords.

**Review by alyapunov — APPROVED** [review]

**Pull request merged**